src/util/psbt/mod.rs

```diff
@@ -196,6 +206,526 @@ impl PartiallySignedTransaction {
+    /// - Err: Error encountered trying to calculate the sighash AND we had the signing key.
+    fn bip32_sign_ecdsa<C, K, T>(
+        &mut self,
+        k: &K,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+        secp: &Secp256k1<C>,
+    ) -> Result<Vec<PublicKey>, SignError>
+    where
+        C: Signing,
+        T: Deref<Target=Transaction>,
+        K: GetKey,
+    {
+        let msg_sighash_ty_res = self.sighash_ecdsa(input_index, cache);
+
+        let input = &mut self.inputs[input_index]; // Index checked in call to `sighash_ecdsa`.
+
+        let mut used = vec![];  // List of pubkeys used to sign the input.
+
+        for (pk, key_source) in input.bip32_derivation.iter() {
+            let sk = if let Ok(Some(sk)) = k.get_key(KeyRequest::Bip32(key_source.clone()), secp) {
+                sk
+            } else if let Ok(Some(sk)) = k.get_key(KeyRequest::Pubkey(PublicKey::new(*pk)), secp) {
+                sk
+            } else {
+                continue;
+            };
+
+            // Only return the error if we have a secret key to sign this input.
+            let (msg, sighash_ty) = match msg_sighash_ty_res {
+                Err(e) => return Err(e),
+                Ok((msg, sighash_ty)) => (msg, sighash_ty),
+            };
+
+            let sig = EcdsaSig {
+                sig: secp.sign_ecdsa(&msg, &sk.inner),
+                hash_ty: sighash_ty,
+            };
+
+            let pk = sk.public_key(secp);
+
+            input.partial_sigs.insert(pk, sig);
+            used.push(pk);
+        }
+
+        Ok(used)
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    ///
+    /// Uses the [`EcdsaSighashType`] from this input if one is specified. If no sighash type is
+    /// specified uses [`EcdsaSighashType::All`]. This function does not support scripts that
+    /// contain `OP_CODESEPARATOR`.
+    pub fn sighash_ecdsa<T: Deref<Target=Transaction>>(
+        &self,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+    ) -> Result<(Message, EcdsaSighashType), SignError> {
+        if self.signing_algorithm(input_index)? != SigningAlgorithm::Ecdsa {
+            return Err(SignError::WrongSigningAlgorithm);
+        }
+
+        let input = self.checked_input(input_index)?;
+        let utxo = self.spend_utxo(input_index)?;
+        let spk = &utxo.script_pubkey; // scriptPubkey for input spend utxo.
+
+        let hash_ty = input.ecdsa_hash_ty()
+            .map_err(|_| SignError::InvalidSighashType)?; // Only support standard sighash types.
+
+        let output_type = self.output_type(input_index)?;
+
+        let is_wpkh = output_type == OutputType::Wpkh;
+        let is_nested_wpkh = output_type == OutputType::ShWpkh;
+
+        let sighash = if output_type.is_segwitv0() {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
```

Comment:
In 303f6ce48460ef15acc17eff32f08f7806a2bf73:

Could you refactor this whole `if` block to match on `output_type`? This would avoid the nested `if output is wsh or wpkh or nested_wpkh { if output is wpkh or nested_wpkh { if output is wpkh {` logic and also make me feel more comfortable about having `OutputType` being what it is.

Reply:
Sure thing, I'll have a play with it. Thanks.

Reply:
Its heaps cleaner. Also, shows up something strange, we have `OutputType::Tr` using `cache.legacy_signature_hash`. Its Friday afternoon here, I've not put thought into whether this is a bug or just misnamed code after taproot was added.
```rust
        let sighash = match self.output_type(input_index)? {
            Bare | Sh | Tr => {
                let script_code = if spk.is_p2sh() {
                    input.redeem_script.as_ref().ok_or(SignError::MissingRedeemScript)?
                } else {
                    spk
                };
                cache.legacy_signature_hash(input_index, script_code, hash_ty.to_u32())?
            },
            Wpkh => {
                let script_code = Script::p2wpkh_script_code(spk).ok_or(SignError::NotWpkh)?;
                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
            }
            ShWpkh => {
                let script_code = Script::p2wpkh_script_code(input.redeem_script.as_ref().expect("checked above"))
                    .ok_or(SignError::NotWpkh)?;
                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
            },
            Wsh | ShWsh => {
                let script_code = input.witness_script.as_ref().ok_or(SignError::MissingWitnessScript)?;
                cache.segwit_signature_hash(input_index, script_code, utxo.value, hash_ty)?
            },
        };
```